File: Leetcode/_2_Add_Two_Numbers.java
// link - https://leetcode.com/problems/add-two-numbers/

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class _2_Add_Two_Numbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        // (probably) Optimal approach - Time = O(n) - Space = O(n)

        ListNode head = new ListNode();
        ListNode curr = head;

        curr.val = (l1.val + l2.val) % 10;
        int carry = (l1.val + l2.val) / 10;

        l1 = l1.next;
        l2 = l2.next;

        while (l1 != null || l2 != null) {
            int n1 = l1 != null ? l1.val : 0;
            int n2 = l2 != null ? l2.val : 0;
            int sum = n1 + n2 + carry;

            carry = sum / 10;

            ListNode temp = new ListNode(sum % 10);
            curr.next = temp;
            curr = temp;

            if (l1 != null)
                l1 = l1.next;
            if (l2 != null)
                l2 = l2.next;
        }

        if (carry > 0) {
            ListNode temp = new ListNode(carry);
            curr.next = temp;
            curr = temp;
        }

        return head;
    }
}

File: Leetcode/python/_2_Add_Two_Numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional


class Solution:
    # Cleaner Version - Time = O(n) - Space = O(1)
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        cur = dummy
        carry = 0

        while l1 or l2 or carry:
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0

            total = v1 + v2 + carry
            carry = 0 if total < 10 else 1
        
            cur.next = ListNode(total if total < 10 else total % 10)
            cur = cur.next

            # Safely advance l1 and l2
            if l1: l1 = l1.next
            if l2: l2 = l2.next
        
        return dummy.next
    
    # Simulate addition by hand, the units are already positioned to be added correctly, just carry the "carry"
    # from left to right instead of right to left
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        head = ListNode()
        res = head
        i, j = l1, l2
        while i and j:
            v = i.val + j.val + carry
            if carry == 1:
                carry = 0

            if v < 10:
                res.next = ListNode(v)
            else:
                carry = 1
                res.next = ListNode(v % 10)

            res = res.next
            i = i.next
            j = j.next

        i = i if i else j
        while i:
            v = i.val + carry
            if carry == 1:
                carry = 0

            if v < 10:
                res.next = ListNode(v)
            else:
                carry = 1
                res.next = ListNode(v % 10)
            res = res.next
            i = i.next

        if carry:
            res.next = ListNode(1)
            res = res.next

        return head.next
